Support read_concern in database.get_collection()

m-aciek opened the pull request:
This argument is passed to underlying methods, so effectively is supported, only check removal is required.

m-aciek committed:
Support read_concern in database.get_collection()

Remove test checking that read_concern is not supported.

m-aciek commented:
Read concern is supported since 0.19.1 #628

m-aciek commented:
@pcorpet would it be possible for you to find time to review and merge this change?

pcorpet commented:
Hi @m-aciek and thanks for your contribution. If we're going to go that way, please make it possible to use a `with_options` (both in collection and database).

However two things to do before removing those checks:
- what should change in the behavior with a different read concern? From a quick glimpse to the doc, it seems that with a in-memory implementation with only one replica, all the read concerns will have the same effect. Please confirm.
- let's still make sure that setting a bad type as a read_concern param in any of those functions would return the same error that `pymongo` would raise.

Commit:
Support read_concern in database.with_options()

Commit:
Support read_concern in Database constructor

Throw a TypeError for read concern of wrong type.

m-aciek commented:
Thanks for looking into this.

Read concern works already in `with_options` of collection.

I think read concern shouldn't make a difference when using one replica, which we are mocking.

I wonder if I should bring a duck typing like in collection.with_options() (with using _WITH_OPTIONS_KWARGS) or just checking the type (first trying to import it from PyMongo, then from mongomock) is enough?

pcorpet commented:
You're right, I think we're fine like this.

Pease cover the new lines with some new tests, and I'll merge it.

Commit:
Support read_concern type check in Collection constructor

Throw a TypeError for read concern of wrong type. Move check from with_options() to underlying constructor.

m-aciek commented:
I've added missing test. I've added read concern type check support in Collection constructor (the constructor may be called in Database().get_collection() method). Also removed read concern type check from Collection().with_options() as it is now handled in the constructor.

m-aciek commented:
Pylint failure is not connected with changes here. I've added a new issue to track that: #709.

Is there a way to retrigger Codecov check?

pcorpet committed:
Fix inconsistant quotes

m-aciek commented:
Added test for database.read_concern attribute, added setting default ReadConcern() if no value provided.

pcorpet commented:
Thanks for all the cleanups and tests.
